packages/financial-templates-lib/test/price-feed/LPPriceFeed.js

```diff
@@ -0,0 +1,152 @@
+  let mockTime = 0;
+  let dummyLogger;
+  let poolDecimals = 18;
+  let tokenDecimals = 8;
+  let priceFeedDecimals = 12;
+
+  beforeEach(async function() {
+    pool = await ERC20.new("LP Pool", "LP", poolDecimals, { from: owner });
+    token = await ERC20.new("Test Token", "TT", tokenDecimals, { from: owner });
+    pool.addMinter(owner);
+    token.addMinter(owner);
+
+    dummyLogger = winston.createLogger({
+      level: "info",
+      transports: [new winston.transports.Console()]
+    });
+
+    lpPriceFeed = new LPPriceFeed({
+      logger: dummyLogger,
+      web3,
+      getTime: () => mockTime,
+      erc20Abi: ERC20Interface.abi,
+      tokenAddress: token.address,
+      poolAddress: pool.address,
+      priceFeedDecimals
+    });
+  });
+
+  it("Basic current price", async function() {
+    await pool.mint(owner, parseFixed("100", poolDecimals));
+    await token.mint(pool.address, parseFixed("25", tokenDecimals));
+    await lpPriceFeed.update();
+
+    assert.equal(lpPriceFeed.getCurrentPrice().toString(), parseFixed("0.25", priceFeedDecimals).toString());
+  });
+
+  it("Correctly selects most recent price", async function() {
+    await pool.mint(owner, parseFixed("100", poolDecimals));
+    await token.mint(pool.address, parseFixed("25", tokenDecimals));
+    await pool.mint(owner, parseFixed("100", poolDecimals));
+    await lpPriceFeed.update();
+
+    assert.equal(lpPriceFeed.getCurrentPrice().toString(), parseFixed("0.125", priceFeedDecimals).toString());
+  });
+
+  it("Historical Price", async function() {
+    await pool.mint(owner, parseFixed("100", poolDecimals));
+    await token.mint(pool.address, parseFixed("25", tokenDecimals));
+
+    await lpPriceFeed.update();
+
+    // Ensure that the next block is mined at a later time.
+    const { timestamp: firstPriceTimestamp } = await web3.eth.getBlock("latest");
+    await advanceBlockAndSetTime(web3, firstPriceTimestamp + 10);
+
+    await pool.mint(owner, parseFixed("100", poolDecimals));
+
+    const { timestamp: secondPriceTimestamp } = await web3.eth.getBlock("latest");
+
+    assert.equal(
+      (await lpPriceFeed.getHistoricalPrice(firstPriceTimestamp)).toString(),
+      parseFixed("0.25", priceFeedDecimals)
+    );
+    assert.equal(
+      (await lpPriceFeed.getHistoricalPrice(firstPriceTimestamp + 5)).toString(),
+      parseFixed("0.25", priceFeedDecimals)
+    );
+    assert.equal(
+      (await lpPriceFeed.getHistoricalPrice(secondPriceTimestamp - 1)).toString(),
+      parseFixed("0.25", priceFeedDecimals)
+    );
+    assert.equal(
+      (await lpPriceFeed.getHistoricalPrice(secondPriceTimestamp)).toString(),
+      parseFixed("0.125", priceFeedDecimals)
+    );
+  });
+
+  it("Zero LP shares", async function() {
+    await token.mint(pool.address, parseFixed("25", tokenDecimals));
+
+    await lpPriceFeed.update();
+
+    assert.equal(lpPriceFeed.getCurrentPrice().toString(), "0");
+  });
+
+  it("Update Frequency", async function() {
```

Comment:
Wonder if this rate limit logic should be a util

Reply:
Agreed. We repeat it in every price feed.